### PR TITLE
Add CLI command to totally delete a DAG

### PR DESCRIPTION
**Pull request by jlowin**
By popular demand: #1341, #1314, #1302, others...

### CODE REVIEW

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -607,6 +615,11 @@ class CLIFactory(object):
             ("--stdout", ), "Redirect stdout to this file"),
         'log_file': Arg(
             ("-l", "--log-file"), "Location of the log file"),
+        'yes': Arg(
+            ("-y", "--yes"),
+            "Do not prompt to confirm. Use with care!",
+            "store_true",
```
Comment (r39132) [inline airflow/bin/cli.py:621]: What does store_true mean?

Comment (artwr) [inline airflow/bin/cli.py:621]: it means that you do not need to specify a value after the flag, the presence of the CLI flag will set the associated value to `True`

File: airflow/models.py [hunk 2]
```diff
@@ -2805,6 +2805,32 @@ def cli(self):
         args = parser.parse_args()
         args.func(args, self)
 
+    @provide_session
+    def delete(self, confirm_prompt=True, session=None):
+        """
+        Deletes the DAG and all related information from the database
+        """
+        if confirm_prompt:
```
Comment (bolkedebruin) [inline airflow/models.py:2813]: I think this is a really weird place for a confirmation prompt

File: airflow/models.py [hunk 3]
```diff
@@ -2805,6 +2805,32 @@ def cli(self):
...
+        session.query(XCom).filter(XCom.dag_id == self.dag_id).delete()
+        session.query(SlaMiss).filter(SlaMiss.dag_id == self.dag_id).delete()
+        session.query(TaskInstance).filter(
+            TaskInstance.dag_id == self.dag_id).delete()
+        session.query(Log).filter(Log.dag_id == self.dag_id).delete()
+        session.query(BaseJob).filter(BaseJob.dag_id == self.dag_id).delete()
+        session.query(DagRun).filter(DagRun.dag_id == self.dag_id).delete()
+        session.query(DagModel).filter(DagModel.dag_id == self.dag_id).delete()
+        session.commit()
+
```
Comment (r39132) [inline airflow/models.py:2833]: We also need to move the files so that they don't keep getting reimported

**Comment by jlowin** [comment]
@bolkedebruin @r39132 (feel free to edit comment)

Summarizing our conversation and open questions:
1. (agreed) it is useful to have a way to delete DAGs from the DB
2. However, they will be reimported almost immediately, so this actually isn't as much of a "delete" as a "clear".
--> action: rename CLI command
3. One solution is as follows : when we delete a record from the DAG table, we also add a tombstone for the DAG in a dag_tombstone table. Any process that reloads the DAG checks the tombstone table and moves it to the local deleted_dags folder. Tombstones are expired after a configurable period of time (e.g. 1 day/week/month) so that the tombstone table does not keep growing unchecked. If a user wants to undelete a dag, we will remove the tombstone entry and move the file back to the dags folder from the delete_dags folder. This will serve use-cases where users wanted to do a refresh.
3. a. Store the hash of the dagfile in the tombstone table. That way users could update the dagfile and not have it be considered the same deleted dag.
4. Better approach, perhaps: use API to delete? @bolkedebruin will need to expand on this.

File: airflow/bin/cli.py (hunk 1 above)
Comment (r39132) [inline airflow/bin/cli.py:621]: Thx @artwr

**Comment by jlowin** [comment]
@r39132 I think we touched on this but filesystem manipulation is no good. For example, each of our Airflow envs' `dag_folders` are synced via `git pull` every 30 seconds. All local changes are discarded. If we use the tombstone table + hash, we could achieve this without moving files.

**Comment by LanceNorskog** [comment]
What if you have to remove the DAG file out yourself before running the CLI command?
It could have another option: "show list of obsolete DAGs".

We keep a separate directory of obsolete jobs just to archive them. I have no problem saying 
"git mv dags/job.py archived/job.py"
before running the database purger.

**Comment by msumit** [comment]
Hi @jlowin .. Sorry for delay in reviewing this PR. If you still wants to pursue this PR, you've to do following things:
1. Create a JIRA under Apache Airflow project (https://issues.apache.org/jira/browse/AIRFLOW/)
2. Rebase you branch with master and resolve conflicts
3. Commit message should follow the guidelines, especially the last section.
4. Add some basic specs and make sure travis stays green.

We've 1 weeks time before this PR gets marked as Abandoned and finally gets closed without being merged to master.

File: airflow/models.py (hunk 2 above)
Comment (tolsac) [reply airflow/models.py:2813]: totally agree

File: airflow/bin/cli.py (hunk 1 above)
Comment (dlamblin) [reply airflow/bin/cli.py:621]: Current style seems to put lines 619-621 on one line.

**Pull request closed without merging**